### ISSUE TITLE
The server returned an empty string for GET

### ISSUE THREAD

**Issue by ArtOfSettling** [ember-cli/ember-ajax#161]
Our server is not something we can modify, and seems to return an empty string for a get with a certain EndPoint. Whilst this only shows in the console as an error, an exception is thrown which causes execution to skip through our catch block.

Everything works ok, the backend performs the work and returns as it always does, however I need a better way to handle and parse this error. Any suggestions?

      this.get('ajax').request('/users/restorepassword/' + this.get('username'), {})
        .then(function() {
          _this.set('infoMessage', 'Please check your email!');
          _this.set('errorMessage', '');
        })
        .catch(function(error) {
          // Execution reaches here, everything is ok, the server returns 200 (with an empty string)
          _this.set('infoMessage', '');
          _this.set('errorMessage', error);
        });

**Comment by alexlafroscia** [comment]
Interesting.  I'll try to determine why an empty string is interpreted as an error and see if we can fix that.

In the meantime, I would need a stack trace of the error in order to direct you to the correct methods to overwrite/modify.

**Comment by ArtOfSettling** [comment]
I'll fire it up here when I get a chance, in the mean time, you can spoof it pretty easily using ember-pretender. (Both my test and runtime code fail with the same error).


  var server = new Pretender(function() {
    this.get("/restorepassword/:name", function() {
      return [
        200,
        {'content-type': 'application/javascript'},
        "" // <------- This being the key
      ];
    });
  });

  server.unhandledRequest = function() {
    console.log("Unhandled Request");
  }

  visit('/resetpassword');
  fillIn('#username', 'Terry');

  click('button:contains(Reset)');

  return wait().then(function() {
    assert.equal(find('.alert-info').length, 1, 'should see info after a correct password restore');
  });

**Comment by alexlafroscia** [comment]
Right, that's what I was planning to do to go about fixing it; create a regression test with that setup and work from there.  Thanks for confirming that that will reproduce the problem, though!

**Comment by ArtOfSettling** [comment]
Yeah, sorry that I couldn't be of more help, I'm not at my main machine right now, if I get back there before you get back to me here, I'll post a call stack ;)

**Comment by alexlafroscia** [comment]
Sure thing. I'll try to take a look at this in the next few days; as long as I can reproduce in a test, then I don't need the call stack.

**Comment by ArtOfSettling** [comment]
Hey @alexlafroscia

Any movement on this?

- T

**Comment by alexlafroscia** [comment]
No, sorry.  Totally lost track of this, it's slipped my mind.  I'll try to take a look soon, but would greatly appreciate a PR if you have the time yourself!

**Comment by alexlafroscia** [comment]
So, I looked into this more.  I'm _pretty_ sure that the error is actually down in jQuery, not `ember-ajax`.  I made a failing test, but the stack trace lands somewhere in jQuery land.

Can you verify that making a request to that stubbed endpoint (just returning a strong) work in jQuery without `ember-ajax`?

**Comment by ArtOfSettling** [comment]
Hey! Sorry for not sorting you out with a PR, I've been a bit too busy to look back into this issue. To somewhat (but not really) answer your question, I was using Ember AJAX prior to using pretender and everything worked well.

Ember.$.ajax({
        type: "GET",
        url: ' https://api.myapp/1/users/restorepassword/SausageMuncher' ,
        processData: false,
        contentType: "application/json; charset=UTF-8",
        headers: {
          'application-id': 'xxx,
          'secret-key': 'xxx'
        }
      }).then(function() {
        Ember.run(function() {
          _this.set('infoMessage', 'Please check your email!');
          _this.set('errorMessage', '');
        });
      }, function(xhr) {
        Ember.run(function() {
          var messageObj = JSON.parse(xhr.responseText);
          _this.set('infoMessage', '');
          _this.set('errorMessage', messageObj.message);
        });
      });


I appreciate there is not much error handling in here.

**Comment by alexlafroscia** [comment]
>  I was using Ember AJAX prior to using pretender and everything worked well.

Subtle difference here; `Ember.$.ajax` is _not_ Ember AJAX, it's just jQuery AJAX.  If the code above works, then the issue must be somewhere in `ember-ajax` after all.  I'll look into it again tonight.

**Comment by XaserAcheron** [comment]
I didn't dive too far into this one, but I've pushed a failing test to my fork so there's something to muck with.

The exact error message I'm getting is `"Unexpected end of JSON input"` -- likely somewhere within jQuery there's an uncaught call to JSON.Parse on the empty string, which will produce that exact error.

Looking at @ArtOfSettling's code above, I'm thinking the reason raw `Ember.$.ajax` is working just fine is that the error handling function in the `.then` call (which is jqXHR's `.then`, not Ember's) is catching the parse error, whereas `ember-ajax` is letting it bubble up. From a quick scan of the code, I'm not seeing anything in `ember-ajax`'s `_makeRequest` that handles outright _failure_ of the ajax request (e.g. parse errors and whatnot). Perhaps there's a deeper discussion to be had here?

**Comment by ezekg** [comment]
I'm seeing this issue as well. API is responding with 202 no content but it's still trying to parse the response as JSON. Having to do this as a workaround:

```javascript
this.get('ajax')
  .post(`https://api.example.com/foo/bar`, {
    contentType: 'application/json',
    data: JSON.stringify({ data })
  })
  .catch((e) => {
    if (e.message === "Unexpected end of JSON input") {
      return RSVP.resolve()
    } else {
      return RSVP.reject(e)
    }
  })
  .then(() => {
    // …
  })
  .catch((e) => {
    // …
  })
```

**Comment by thisFunction** [comment]
@ezekg, I had the same problem and adding `dataType: 'text'` to post method fixed the error

**Comment by mukk85** [comment]
Thank you @thisFunction that solved it for me as well.

**Comment by GitVitor** [comment]
Nothing yet? I got this problem too

**Comment by alexlafroscia** [comment]
Always up for reviewing a PR!
